torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -60,88 +62,176 @@ def reset_table(self) -> None:
+    import functools
+
+    import triton
+    from triton.compiler.code_generator import ast_to_ttir
+    from triton.runtime.jit import JITFunction
+
+    assert isinstance(kernel, JITFunction)
+
+    args = [val for key, val in kwargs.items()]
+
+    specialization = kernel._get_config(*args)
+    constants = {i: arg for i, arg in enumerate(args) if not isinstance(arg, Tensor)}
+    debug = None
+    target = None
+
+    # Build kernel signature -- doesn't include constexpr arguments.
+    signature = {
+        i: kernel._type_of(kernel._key_of(arg))
+        for i, arg in enumerate(args)
+        if i not in kernel.constexprs
     }
 
-    def __init__(self, infos) -> None:
-        super().__init__()
-        self.infos = infos
-        self.read_depth = 0
-        self.in_store = False
+    @dataclasses.dataclass(frozen=True)
+    class TensorParam:
+        idx: int
+
+    @dataclasses.dataclass(frozen=True)
+    class NonTensorParam:
+        source: Any
+        pass
+
+    @dataclasses.dataclass(frozen=True)
+    class Intermediate:
+        idx: int
+
+    mappings: Dict[Any, Union[TensorParam, NonTensorParam, Intermediate]] = dict()
+    next_intermediate = 0
+
+    def convert(arg):
+        if isinstance(arg, triton._C.libtriton.triton.ir.block_argument):
+            if arg not in mappings:
+                mappings[arg] = NonTensorParam(arg)
+            return mappings[arg]
+        if isinstance(arg, triton._C.libtriton.triton.ir.value):
+            if arg not in mappings:
+                nonlocal next_intermediate
+                mappings[arg] = Intermediate(next_intermediate)
+                next_intermediate += 1
+            return mappings[arg]
+        return arg
+
+    # Name of mutation op to mutated parameter indices
+    MUTATION_OPS = {"masked_store": [0]}
+
+    ops: Dict[Union[TensorParam, NonTensorParam, Intermediate], "Op"] = dict()
+    sinks: List["Op"] = []
+
+    @dataclasses.dataclass
+    class Op:
+        name: str
+        args: List[Any]
+        kwargs: Dict[str, Any]
+
+        @staticmethod
+        def create(name, args, kwargs, result):
+            op = Op(name, [convert(a) for a in args], kwargs)
+            ops[convert(result)] = op
+            if name in MUTATION_OPS:
+                sinks.append(op)
+
+    class MutationAnalysisWrapper:
+        def __init__(self, builder):
+            self.builder = builder
+
+        def __getattr__(self, name):
+            def generic_indirection(name, *args, **kwargs):
+                result = getattr(self.builder, name)(*args, **kwargs)
+
+                if name.startswith("create_"):
+                    Op.create(name[len("create_") :], args, kwargs, result)
+                elif name in {"get_int32"}:
+                    Op.create(name[len("get_") :], args, kwargs, result)
+                elif name in {"get_loc", "set_loc"}:
+                    pass
+                else:
+                    pass
+                return result
```

Comment:
There isn't a 1:1 mapping between python wrapper objects and underlying C++ objects, a new wrapper is created every time it's returned from C++. This means your dict lookup will never match. You need to create your own python wrapper object and return those from the builder, not the original pybind11 wrapper objects.

Reply:
For operations over non block arguments, it seems to always be the same pointer but it does not match for block argument.

I don't think I am following the new wrapper train of thought.
If I were to create
```
class CustomIRBlockArgument(triton._C.libtriton.triton.ir.block_argument):
   ...
```
and do `if isinstance(result, triton._C.libtriton.triton.ir.block_argument)`, wrap it with `CustomIRBlockArgument`. What difference would that make?

I have also tried
```
triton._C.libtriton.triton.ir.block_argument = CustomIRBlockArgument
```
so that I can overwrite it but that did not work, i assume something around pybind issues.

Reply:
Wondering if `pybind11` wrapper object is opaque to us, or we could extract the pointer (or other identifier) of the underlying C++ object which doesn't change? Unless this sounds too low-level.

Reply:
> I don't think I am following the new wrapper train of thought.

If I were implementing this I would use composition
```python
@dataclass
class CustomFuncOp:
     underlying: triton._C.libtriton.triton.ir.function
     ...
```

Create wrappers like these for each of the types defined in triton's `ir.cc` file. Then we have complete control over the interface without needing to monkey patch anything. All the state we care about stays entirely in python and potentially you could forego the `underlying` part entirely and just stub out everything that's not relevant to the analysis.

> I guess i am still not sure why we did not do the custom MLIR parser since that was much more trivial than this but I'm willing to see this approach through.

It seems to me that with this method skip parsing and goes directly to traversing the tree of operations which is what we actually care about. It's possible my estimation is off though and perhaps the representation returned by a parser would be much easier to work with.